### PR TITLE
modules/system/defaults/dock.nix: add options for hot corners

### PR DESCRIPTION
**Pull request by tboerger**

### CODE REVIEW

**Comment by malob** [comment]
Cool! FWIW, I think it would be helpful if the options descriptions explained what `int` values correspond to which hot corner behaviors.

**Comment by tboerger** [comment]
Makes sense, do you have a suggestion how to format the description? These are the available values:

0: no-op
2: Mission Control
3: Show application windows
4: Desktop
5: Start screen saver
6: Disable screen saver
7: Dashboard
10: Put display to sleep
11: Launchpad
12: Notification Center

File: modules/system/defaults/dock.nix [hunk 1]
```diff
@@ -162,5 +162,37 @@ in {
       '';
     };
 
+    system.defaults.dock.wvous-tl-corner = mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      description = ''
+        Hot corner action for top left corner.  The default is null.
```
Comment (malob) [inline modules/system/defaults/dock.nix:169]:
```suggestion
        Hot corner action for top left corner. Valid values include:

        <itemizedlist>
        <listitem><para><literal>0</literal>: No-op</para></listitem>
        <listitem><para><literal>2</literal>: Mission Control</para></listitem>
        <listitem><para><literal>3</literal>: Show application windows</para></listitem>
        <listitem><para><literal>4</literal>: Desktop</para></listitem>
        <listitem><para><literal>5</literal>: Start screen saver</para></listitem>
        <listitem><para><literal>6</literal>: Disable screen saver</para></listitem>
        <listitem><para><literal>7</literal>: Dashboard</para></listitem>
        <listitem><para><literal>10</literal>: Put display to sleep</para></listitem>
        <listitem><para><literal>11</literal>: Launchpad</para></listitem>
        <listitem><para><literal>12</literal>: Notification Center</para></listitem>
        </itemizedlist>
```

File: modules/system/defaults/dock.nix [hunk 2]
```diff
@@ -162,5 +162,37 @@ in {
...
+      '';
+    };
+
+    system.defaults.dock.wvous-bl-corner = mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      description = ''
+        Hot corner action for bottom left corner.  The default is null.
```
Comment (malob) [inline modules/system/defaults/dock.nix:177]:
```suggestion
        Hot corner action for bottom left corner. Valid values include:

        <itemizedlist>
        <listitem><para><literal>0</literal>: No-op</para></listitem>
        <listitem><para><literal>2</literal>: Mission Control</para></listitem>
        <listitem><para><literal>3</literal>: Show application windows</para></listitem>
        <listitem><para><literal>4</literal>: Desktop</para></listitem>
        <listitem><para><literal>5</literal>: Start screen saver</para></listitem>
        <listitem><para><literal>6</literal>: Disable screen saver</para></listitem>
        <listitem><para><literal>7</literal>: Dashboard</para></listitem>
        <listitem><para><literal>10</literal>: Put display to sleep</para></listitem>
        <listitem><para><literal>11</literal>: Launchpad</para></listitem>
        <listitem><para><literal>12</literal>: Notification Center</para></listitem>
        </itemizedlist>
```

File: modules/system/defaults/dock.nix [hunk 3]
```diff
@@ -162,5 +162,37 @@ in {
...
+      '';
+    };
+
+    system.defaults.dock.wvous-tr-corner = mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      description = ''
+        Hot corner action for top right corner.  The default is null.
```
Comment (malob) [inline modules/system/defaults/dock.nix:185]:
```suggestion
        Hot corner action for top right corner. Valid values include:

        <itemizedlist>
        <listitem><para><literal>0</literal>: No-op</para></listitem>
        <listitem><para><literal>2</literal>: Mission Control</para></listitem>
        <listitem><para><literal>3</literal>: Show application windows</para></listitem>
        <listitem><para><literal>4</literal>: Desktop</para></listitem>
        <listitem><para><literal>5</literal>: Start screen saver</para></listitem>
        <listitem><para><literal>6</literal>: Disable screen saver</para></listitem>
        <listitem><para><literal>7</literal>: Dashboard</para></listitem>
        <listitem><para><literal>10</literal>: Put display to sleep</para></listitem>
        <listitem><para><literal>11</literal>: Launchpad</para></listitem>
        <listitem><para><literal>12</literal>: Notification Center</para></listitem>
        </itemizedlist>
```

File: modules/system/defaults/dock.nix [hunk 4]
```diff
@@ -162,5 +162,37 @@ in {
...
+      '';
+    };
+
+    system.defaults.dock.wvous-br-corner = mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      description = ''
+        Hot corner action for bottom right corner.  The default is null.
```
Comment (malob) [inline modules/system/defaults/dock.nix:193]:
```suggestion
        Hot corner action for bottom right corner. Valid values include:

        <itemizedlist>
        <listitem><para><literal>0</literal>: No-op</para></listitem>
        <listitem><para><literal>2</literal>: Mission Control</para></listitem>
        <listitem><para><literal>3</literal>: Show application windows</para></listitem>
        <listitem><para><literal>4</literal>: Desktop</para></listitem>
        <listitem><para><literal>5</literal>: Start screen saver</para></listitem>
        <listitem><para><literal>6</literal>: Disable screen saver</para></listitem>
        <listitem><para><literal>7</literal>: Dashboard</para></listitem>
        <listitem><para><literal>10</literal>: Put display to sleep</para></listitem>
        <listitem><para><literal>11</literal>: Launchpad</para></listitem>
        <listitem><para><literal>12</literal>: Notification Center</para></listitem>
        </itemizedlist>
```

**Review by malob — COMMENTED** [review]
Here's how I'd include the valid values in the description.

**Edit:**
I also used my [`prefmanager`](https://github.com/malob/prefmanager) tool to double check the values and noticed there were a couple missing, which I added. (I'm on macOS Monterey.)

I also noticed that there seems to be related key, `wvous-{tl,tr,bl,br}-modifier`, that changes sometimes when I change the settings in the UI. For example, when disabling the bottom right hot corner, i.e., setting it to `–`, I see the following change:
```
  wvous-br-modifier (Value changed)
    <integer>1048576</integer>
    <integer>0</integer>
```
(The top value is the old value, the bottom one is the new value.)

Any idea what that key is for? I'm not sure setting it matters, but seems worth bringing to your attention.

**Comment by malob** [comment]
Separately, and this really isn't important, being that I'm a stickler for using precise types when possible, I'd change `types.int` to `types.ints.positive`.

**Comment by tboerger** [comment]
The modifier value is AFAIK used to combine the action with a modifier key, but I never tried that in combination. Should this also be added? Changing the type to be more specific totally makes sense.

**Comment by malob** [comment]
> The modifier value is AFAIK used to combine the action with a modifier key, but I never tried that in combination. Should this also be added?

¯\_(ツ)_/¯

If it were me, I'd test switching between a value of `1` (Disabled) and something else, and if everything works, I'd ignore it. (If in the future, someone does run into an issue where it matters, a future PR could address that.)

**Comment by tboerger** [comment]
I have applied the requested change related to `types.int.positive`. Within my project I have created two additional options for Dock, should I add a commit to this PR? Or should I create a dedicated PR?

```
    system.defaults.dock.largesize = mkOption {
      type = types.nullOr types.int;
      default = null;
      description = ''
        Large size of the icons in the dock.  The default is 64.
      '';
    };

    system.defaults.dock.magnification = mkOption {
      type = types.nullOr types.bool;
      default = null;
      description = ''
        Enable magnification for the icons in the dock.  The default is false.
      '';
    };
```

**Comment by malob** [comment]
Seems fine to me to have one PR for adding a bunch of keys that are all related to the `com.apple.dock` domain (though for the avoidance of doubt, I'm just some guy who likes `nix-darwin` and tries to be helpful, but I don't have any actual authority here 😛).

If you do go ahead and add them, you'll want to change the type of `system.defaults.dock.largesize` from `types.int` to `float` (which is a custom type defined in the module). Also, some (unsolicited) nits:
* I'd use the `<literal></literal>` tag around "64" and "false" in the two descriptions.
* I'd change the first sentence of the description for `system.defaults.dock.largesize` to make it clear that the value is the size the icons will be magnified to if magnification is enabled.

**Comment by tboerger** [comment]
I have rebased my commits and applied the correct change for `types.ints.positive`. Would be cool to get a review/approve :)

**Review by malob — APPROVED** [review]
LGTM.

**Review by LnL7 — APPROVED** [review]
Thanks!

As a general thought, especially for these kind of more cryptic values, I think it might make sense to introduce modules with more user friendly options (enum, etc.) that internally translates to the corresponding defaults.

**Pull request merged**